# Task #1 of Deep Learning Course- MIT - ICESI 

1. Load libraies

In [205]:
from sklearn.datasets import load_digits
from sklearn.metrics import confusion_matrix
import pylab as pl

2. Load digits

In [206]:
digits = load_digits()

# 7

3. Print some sevens

In [207]:
targets = digits.target 
indices = [i for i,x in enumerate(targets) if x == 7] #identifying seven indices
#indices

pl.gray() # Queremos las imágenes en grises
for i in indices[10:40]:
    #print(digits.images[i])  # Imprimir una imagen dada
    #pl.gray() # Queremos las imágenes en grises
    #pl.matshow(digits.images[i]) # Imprimir una imagen dada
    pass

<Figure size 432x288 with 0 Axes>

It seems like there are two types of 7s: 1. The above line is straight, and 2. The above line is ascendent.

There are some patterns identified: 
    1. The first column is black (number less than or equal to 3). 
    2. The lower-right corner (positions [5][7], [6][7], [7][7], [7][6], [7][5], [6][6]) is black (number less than or equal to 3). 
    3.1. For the first type of 7, almost always there is a black (number less than or equal to 3) line in the 3rd row until the 3rd column (pos [2][2]). 
    3.2. For the second type of 7, it has a black pixel (number less than or equal to 3) wether in [2][3] or in [2][4].

In [208]:
def column(matrix, i):
    return [row[i] for row in matrix]

def is_black(lista): 
    return all(number <= 3 for number in lista)

In [209]:
def is_seven(digit):
    #print (digit)
    
    #1. The first column is black (number less than or equal to 3).
    
    first_column = column(digit, 0)
    fcb = is_black(first_column) #first column black
    #print(fcb)

    #2. The lower-right corner (positions [5][7], [6][7], [7][7], [7][6], [7][5], [6][6]) is black (number less than or equal to 3). 

    lr_corner= [digit[5][7], digit[6][7], digit[7][7], digit[7][6], digit[7][5], digit[6][6]]
    lrb= is_black(lr_corner)
    #print(lr_corner)

    #3.1. For the first type of 7, almost always there is a black (number less than or equal to 3) line in the 3rd row until the 3rd column (pos [2][2]). 

    third_row = [digit[2][0], digit[2][1], digit[2][2]]
    trb = is_black(third_row)
    #print(trb)

    #3.2. For the second type of 7, it has a black pixel (number less than or equal to 3) wether in [2][3] or in [2][4].

    alb = digit[2][3] <= 3 or digit[2][4] <= 3  # ascendent line black
    #print(alb)

    #3

    tc = trb or alb # third condition
    #print(tc)

    result= fcb and lrb and tc
    return result
    #print(result)

In [210]:
#is_seven(digits.images[7])

Confusion Matrix

In [211]:
y_true= [digit==7 for digit in digits.target]
y_pred = [is_seven(digit) for digit in digits.images]
#print (y_pred)
tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
print(tp, fp, fn, tn)

167 355 12 1263


In [212]:
acc = (tp+tn)/(tp+fp+fn+tn)
err = (fp+fn)/(tp+fp+fn+tn)
pre = tp/(tp+fp)
rec = tp/(tp+fn)
spe = tn/(tn+fp)

print (acc, err, pre, rec, spe)

0.7957707289927657 0.2042292710072343 0.31992337164750956 0.9329608938547486 0.7805933250927071


<b> Accuracy = 80% and Error rate = 20%: </b> This means that 80% of the numbers all well classified.
<br>
<b> Precision (PPV) = 32%: </b> This means that the proportion of positive results is low. This is because the negative values are considerably more than positive values. 
<br>
<b> Recall (TPR) = 93%: </b> This means that the proportion of actual positives that are correctly identified as such is very high. 
<br>
<b> Specificity (TNR) = 78%: </b> This means that the proportion of actual negatives that are correctly identified as such is high.

# 9

Print some nines

In [213]:
targets = digits.target 
indices = [i for i,x in enumerate(targets) if x == 9] #identifying seven indices
#indices

pl.gray() # Queremos las imágenes en grises
for i in indices:
    #print(digits.images[i])  # Imprimir una imagen dada
    #pl.gray() # Queremos las imágenes en grises
    #pl.matshow(digits.images[i]) # Imprimir una imagen dada
    pass

<Figure size 432x288 with 0 Axes>

There are some patterns identified: 
    1. The first column is black (number less than or equal to 3). 
    2.1. The upper-right corner (positions [0][7], [0][6], [1][7]) is black (number less than or equal to 3). 
    2.2. The lower-right corner (positions [5][7], [6][7], [7][7], [7][6], [7][5], [6][6]) is black (number less than or equal to 3). 
    3.1 There is a black (number less than or equal to 3) line in the 5th row until the 4th column (pos [4][3]).
    3.2 There is a black (number less than or equal to 3) line in the 6th row until the 4th column (pos [5][3]). 

In [214]:
#def is_almost_black(lista): 
 #   return all(number <= 5 for number in lista)

In [215]:
def is_nine(digit):
    #print (digit)
    
    #1. The first column is black (number less than or equal to 3).
    
    first_column = column(digit, 0)
    fcb = is_black(first_column) #first column black
    #print(fcb)

    #2.1. The upper-right corner (positions [0][7], [0][6], [1][7]) is black (number less than or equal to 3). 
    
    ur_corner= [digit[0][7], digit[0][6], digit[1][7]]
    urb= is_black(ur_corner)
    #print(urb)
    
    #2.2. The lower-right corner (positions [5][7], [6][7], [7][7], [7][6], [7][5], [6][6]) is black (number less than or equal to 3). 
    
    lr_corner= [digit[5][7], digit[6][7], digit[7][7], digit[7][6], digit[7][5], digit[6][6]]
    lrb= is_black(lr_corner)
    #print(lrb)
    
    #2 
    
    sc = urb or lrb  #second condition

    #3.1. There is a almost black (number less than or equal to 5) line in the 5th row until the 4th column (pos [4][3]).

    fifth_row = [digit[4][0], digit[4][1], digit[4][2], digit[4][3]]
    frb = is_black(fifth_row)
    #print(frb)

    #3.2. There is a almost black (number less than or equal to 5) line in the 6th row until the 4th column (pos [5][3]). 

    sixth_row = [digit[5][0], digit[5][1], digit[5][2], digit[5][3]]
    srb = is_black(sixth_row)
    #print(srb)

    #3

    tc = frb or srb # third condition
    #print(tc)

    result= fcb and sc and tc
    return result
    #print(result)

In [216]:
#is_nine(digit)

Confusion Matrix

In [217]:
y_true= [digit==9 for digit in digits.target]
y_pred = [is_nine(digit) for digit in digits.images]
#print (y_pred)
tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
print(tp, fp, fn, tn)

156 350 24 1267


In [218]:
acc = (tp+tn)/(tp+fp+fn+tn)
err = (fp+fn)/(tp+fp+fn+tn)
pre = tp/(tp+fp)
rec = tp/(tp+fn)
spe = tn/(tn+fp)

print (acc, err, pre, rec, spe)

0.7918753478018921 0.20812465219810797 0.308300395256917 0.8666666666666667 0.7835497835497836


<b> Accuracy = 79% and Error rate = 21%: </b> This means that 79% of the numbers all well classified.
<br>
<b> Precision (PPV) = 31%: </b> This means that the proportion of positive results is low. This is because the negative values are considerably more than positive values. 
<br>
<b> Recall (TPR) = 87%: </b> This means that the proportion of actual positives that are correctly identified as such is considerably high. 
<br>
<b> Specificity (TNR) = 78%: </b> This means that the proportion of actual negatives that are correctly identified as such is high.